content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,255 @@
+
+```console
+npm run hello
+```
+
+The source code of this step is as follows, and we will use this lightweight example to get privy on how the data flow 
+in electron works: 
+
+{{ include_code("_code-samples/build-a-wallet/js/1_hello.js", language="js") }}
+
+The main parts are a helper function that executes a ledger request and the main function that creates the application 
+window. The main function then calls our helper function and distributes its contents by broadcasting an event that will get picked up by the frontend.
+
+We will use this example to study how to do Inter Process Communication (IPC) in Electron. For reasons of due diligence it 
+has to be mentioned that JavaScript has no real Processes / Threading because follows event-driven paradigm. Nonetheless 
+Electron provides us with two IPC modules called ipcMain and ipcRenderer. We can roughly equate those to the backend
+and the frontend when we think in terms of client-server applications. It works as follows:
+
+1. Create a function that enables the frontend to subscribe to backend events (in 1_preload.js)
+2. Make the function available by preloading it (webPreferences.preload during window creation)
+3. Use that function in the frontend (e.g. 1_renderer.js, loaded in 1_hello.html) to attach a callback that handles frontend updates when the event is dispatched
+4. Dispatch the event from the backend (e.g. appWindow.webContents.send('update-ledger-index', value))
+
+### 2.A. Show Ledger Updates by Polling
+
+**Full code for this step:** [`2_async-poll.js`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/2_async-poll.js).
+
+Our "Hello Ledger" application from Step 1. so far only shows the latest validated ledger sequence at the time when we 
+opened it. So let's take things up a notch and add some dashboard like functionality where our wallet app will keep in 
+sync with the ledger and display the latest specs and stats like a clock that is keeping track on time:
+
+![Screenshot: Step 2, show ledger updates](img/javascript-wallet-2.png)
+
+To accomplish this, we'll need to subscribe to the ledger using this snippet of code: 
+
+<Code snippet>
+
+When you subscribe to the `ledger` stream, anytime there is a new validated ledger your code will be updated...
+
+To get the application running at this stage of development, run the following command:
+
+```console
+npm run async-poll
+```
+
+The code has been refactored so that the main logic now resides in a main() function which delegates the window creation
+to a helper function. We also put the functionality to do the ledger request into a separate function:
+
+{{ include_code("_code-samples/build-a-wallet/js/2_async-poll.js", language="js", lines="8-22" }}
+
+Our main logic is to continuously poll the XRPL using this function in an infinite loop:
+
+{{ include_code("_code-samples/build-a-wallet/js/2_async-poll.js", language="js", lines="57-71" }}
```

Comment:
Syntax errors in these `include_code` macros (missing the closing parentheses) are causing the build to fail.

```suggestion
{{ include_code("_code-samples/build-a-wallet/js/2_async-poll.js", language="js", lines="8-22") }}

Our main logic is to continuously poll the XRPL using this function in an infinite loop:

{{ include_code("_code-samples/build-a-wallet/js/2_async-poll.js", language="js", lines="57-71") }}
```

Also, this is totally optional, but as a best practice I like to use `start_with=` and `end_before=` instead of `lines` because updates to the code sample file can throw off the line numbers, but `start_with`/`end_before` don't have to be updated unless the changes actually touch those sections.

Reply:
Whoops, I corrected that and forgot to commit the changes before vacation week. Done.

Reply:
Will include your starts_with and ends_before suggestion...

Reply:
Okay, "starts_with" and "before_end" messes up the JavaScript code, will stick to "lines" for now